lib/plugins/threaded-signing/plugin.c

```diff
@@ -137,33 +149,75 @@ signing_worker_thread(void *user_data)
   g_cond_signal(&self->cond);
 
   while (self->is_running) {
-    // Wait for a signal, triggered when it is time to sign a hash.
-    g_cond_wait(&self->cond, &self->mutex);
-
-    if (self->plugin_state == THREADED_SIGNING_HAS_HASH_TO_SIGN) {
-      // Copy the |hash_to_sign| to |signature_info| and start signing. In principle, it is now
-      // possible to prepare for a new |hash_to_sign|.
+    if (self->input_buffer_idx > 0) {
+      // Get the oldest hash from the input buffer
+      // Copy the hash to |signature_info| and start signing.
       assert(self->hash_size == self->signature_info->hash_size);
-      memcpy(self->signature_info->hash, self->hash_to_sign, self->hash_size);
-      self->plugin_state = THREADED_SIGNING_SIGNS_HASH;
+      assert(self->signature_info->hash);
+      memcpy(self->signature_info->hash, self->input_buffer[0], self->hash_size);
+
+      // Store the oldest hash temporarily by moving it from the first element of the buffer to the
+      // end and move all other elements in the buffer forward.
+      uint8_t *tmp = self->input_buffer[0];
+      int j = 0;
+      while (self->input_buffer[j + 1] != NULL && j < MAX_BUFFER_LENGTH - 1) {
+        self->input_buffer[j] = self->input_buffer[j + 1];
+        j++;
+      }
+      self->input_buffer[j] = tmp;
+      self->input_buffer_idx--;
 
       // Let the signing operate outside a lock. Otherwise sv_interface_get_signature() is blocked,
       // since variables need to be read under a lock.
       g_mutex_unlock(&self->mutex);
       SignedVideoReturnCode status = openssl_sign_hash(self->signature_info);
-
+      // TODO: Remove sleeps
+      a++;
+      if (a > 3) {
+        sleep(0);
+      } else {
+        sleep(5);
+      }
       g_mutex_lock(&self->mutex);
-      // When successfully done with signing, move |plugin_state| to THREADED_SIGNING_HAS_SIGNATURE,
-      // otherwise move to THREADED_SIGNING_ERROR to report the error when getting the signature.
+
+      // |output_buffer| is full. Buffers this long are not supported.
+      if (self->output_buffer_idx >= MAX_BUFFER_LENGTH) {
+        self->is_running = false;
```

Comment:
Stopping the thread will just stop the thread. Calls to plugin-APIs from the library thread will still keep running. The output buffer will never be filled further, but the input buffer will and the ```threaded_openssl_sign_hash()``` API will keep returning SV_OK until we reach ```MAX_BUFFER_LENGTH```.

The library will also keep a lot of unfinished SEI-frames. How can we make life easier for the main thread to know if something went wrong so proper actions can be made there as well.

You can mimic everything through the test code you have with proper settings and sleep(), free(), printf() and other actions.